src/analyze/analyze-shutdown-blame.c

```diff
@@ -0,0 +1,227 @@
+#include "sort-util.h"
+#include "string-util.h"
+#include "time-util.h"
+
+typedef struct ShutdownTime {
+        char *name;
+        usec_t start_time;
+        usec_t stop_time;
+        usec_t duration;
+} ShutdownTime;
+
+static ShutdownTime* shutdown_time_free(ShutdownTime *t) {
+        if (!t)
+                return NULL;
+
+        free(t->name);
+        return mfree(t);
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(ShutdownTime*, shutdown_time_free);
+
+static int compare_shutdown_times(ShutdownTime * const *a, ShutdownTime * const *b) {
+        if ((*a)->duration > (*b)->duration)
+                return -1;
+        if ((*a)->duration < (*b)->duration)
+                return 1;
+        return 0;
+}
+
+static int acquire_shutdown_times(Hashmap **ret) {
+        _cleanup_(sd_journal_closep) sd_journal *j = NULL;
+        _cleanup_(hashmap_freep) Hashmap *shutdown_times = NULL;
+        const void *data;
+        size_t length;
+        int r;
+
+        shutdown_times = hashmap_new(&string_hash_ops);
+        if (!shutdown_times)
+                return log_oom();
+
+        r = sd_journal_open(&j, SD_JOURNAL_LOCAL_ONLY);
+        if (r < 0)
+                return log_error_errno(r, "Failed to open journal: %m");
+
+        /* Look for systemd messages with units */
+        r = sd_journal_add_match(j, "SYSLOG_IDENTIFIER=systemd", 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to add journal match: %m");
+
+        /* Seek to the end and go backwards to find the most recent shutdown */
+        r = sd_journal_seek_tail(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to seek to journal tail: %m");
+
+        r = sd_journal_previous(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to iterate journal: %m");
+
+        SD_JOURNAL_FOREACH_BACKWARDS(j) {
+                const char *message = NULL, *unit = NULL;
+                usec_t timestamp;
+                _cleanup_(shutdown_time_freep) ShutdownTime *st = NULL;
+                ShutdownTime *existing;
+
+                r = sd_journal_get_realtime_usec(j, &timestamp);
+                if (r < 0)
+                        continue;
+
+                r = sd_journal_get_data(j, "MESSAGE", &data, &length);
+                if (r < 0)
+                        continue;
+
+                message = (const char*) data + STRLEN("MESSAGE=");
+                if (length <= STRLEN("MESSAGE="))
+                        continue;
+
+                r = sd_journal_get_data(j, "UNIT", &data, &length);
+                if (r < 0)
+                        continue;
+
+                unit = (const char*) data + STRLEN("UNIT=");
+                if (length <= STRLEN("UNIT="))
+                        continue;
+
+                /* Skip if this is not a stopping or stopped message */
+                if (!startswith(message, "Stopping") && !startswith(message, "Stopped"))
+                        continue;
```

Comment:
nope, please do no match by english language strings. They can change any time. And who knows maybe get translated one day. They are *not* API.

these messages are already recognizable by their MESSAGE_ID= and that's what they are intended for.

For example, use "journalctl MESSAGE_ID=de5b426a63be47a7b6ac3eaac82e2f6f" to get a list of all Stopping messages. You get the idea.